contracts/randdrop/src/contract.rs

```diff
@@ -0,0 +1,843 @@
+    nois_proxy_denom: Option<String>,
+    nois_proxy_amount: Option<Uint128>,
+    nois_proxy_address: Option<String>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    // check the calling address is the authorised multisig
+    ensure_eq!(info.sender, config.manager, ContractError::Unauthorized);
+
+    let manager = match manager {
+        Some(ma) => deps.api.addr_validate(&ma)?,
+        None => config.manager,
+    };
+    let randdrop_denom = randdrop_denom.unwrap_or(config.randdrop_denom);
+    let nois_proxy_denom = nois_proxy_denom.unwrap_or(config.nois_proxy.price.denom);
+    let nois_proxy_amount = nois_proxy_amount.unwrap_or(config.nois_proxy.price.amount);
+    let nois_proxy_address = match nois_proxy_address {
+        Some(np) => deps.api.addr_validate(&np)?,
+        None => config.nois_proxy.address,
+    };
+
+    let nois_proxy = NoisProxy {
+        price: Coin {
+            denom: nois_proxy_denom,
+            amount: nois_proxy_amount,
+        },
+        address: nois_proxy_address,
+    };
+
+    CONFIG.save(
+        deps.storage,
+        &Config {
+            manager,
+            randdrop_denom,
+            nois_proxy,
+        },
+    )?;
+
+    Ok(Response::new().add_attribute("action", "update_config"))
+}
+
+// This function will call the proxy and ask for the randomness round
+pub fn execute_randdrop(
+    deps: DepsMut,
+    info: MessageInfo,
+    amount: Uint128,
+    proof: Vec<HexBinary>,
+) -> Result<Response, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+
+    if !MERKLE_ROOT.exists(deps.storage) {
+        return Err(ContractError::MerkleRootAbsent);
+    }
+
+    if PARTICIPANTS.has(deps.storage, &info.sender) {
+        return Err(ContractError::RandomnessAlreadyRequested);
+    }
+
+    if CLAIMED.has(deps.storage, &info.sender) {
+        return Err(ContractError::Claimed {});
+    }
+
+    // The contract will spend NOIS tokens on this info.sender to buy randomness.
+    // To prevent spam and users abusing the funds, we only allow addresses listed on the randdrop to call this entrypoint
+    // Sending the proof here will make sure that the sender is a potential randdrop winner
+    let merkle_root = MERKLE_ROOT.load(deps.storage)?; // This can be optimised by not loading the state twice
+    if !is_proof_valid(info.sender.clone(), amount, merkle_root, proof)? {
+        return Err(ContractError::Unauthorized {});
+    }
+
+    // Get the price from the nois-proxy
+
+    let response = Response::new().add_message(WasmMsg::Execute {
+        contract_addr: config.nois_proxy.address.into_string(),
+        // GetNextRandomness requests the randomness from the proxy
+        // The job id is needed to know what randomness we are referring to upon reception in the callback.
+        msg: to_binary(&ProxyExecuteMsg::GetNextRandomness {
+            job_id: "randdrop-".to_string() + info.sender.as_str(),
```

Comment:
```suggestion
            job_id: format!("randdrop-{}", info.sender),
```

this looks a bit nicer but does the same